### PR TITLE
[4.0] JHtml prepared statements

### PR DESCRIPTION
**Pull request by SharkyKZ**
### Summary of Changes

This adds prepared statements and cleans up queries in JHtml classes.

### Testing Instructions

Test that the following methods work like before PR:

```
JHtmlAccess::level()
JHtmlAccess::assetgroups()
JHtmlCategory::options()
JHtmlCategory::categories()
JHtmlContentlanguage::existing()
JHtmlList::users()
JHtmlMenu::menus()
JHtmlMenu::menuItems()
JHtmlMenu::linkOptions()
JHtmlTag::options()
JHtmlTag::tags()
JHtmlUser::userlist()
JHtmlWorkflowstage::existing()
```

### Expected result

Works like before.

### Documentation Changes Required
No.

### CODE REVIEW

File: libraries/cms/html/category.php [hunk 1]
```diff
@@ -81,38 +84,30 @@ public static function options($extension, $config = array('filter.published' =>
...
 
 			// Filter on the access
 			if (isset($config['filter.access']))
 			{
-				if (is_string($config['filter.access']))
+				if (is_numeric($config['filter.access']))
 				{
-					$query->where('a.access = ' . $db->quote($config['filter.access']));
+					$query->where($db->quoteName('a.access') . ' = :access')
+						->bind(':access', $config['filter_access']);
```
Comment (Quy) [inline libraries/cms/html/category.php]:
```suggestion
						->bind(':access', $config['filter_access'], ParameterType::INTEGER);
```

File: libraries/cms/html/menu.php [hunk 2]
```diff
@@ -90,30 +93,32 @@ public static function menuItems($config = array())
...
+				->select($db->quoteName(['a.id', 'a.title', 'a.level', 'a.menutype', 'a.client_id'], ['value', 'text', null, null, null]))
+				->from($db->quoteName('#__menu', 'a'))
+				->where($db->quoteName('a.parent_id') . ' > 0');
 
 			// Filter on the client id
 			if (isset($clientId))
 			{
-				$query->where('a.client_id = ' . (int) $clientId);
+				$query->where($db->quoteName('a.client_id') . ' = :client')
+					->bind(':client', $clientId);
```
Comment (Quy) [inline libraries/cms/html/menu.php]:
```suggestion
					->bind(':client', $clientId, ParameterType::INTEGER);
```

File: libraries/cms/html/menu.php [hunk 3]
```diff
@@ -90,30 +93,32 @@ public static function menuItems($config = array())
...
 				if (is_numeric($config['published']))
 				{
-					$query->where('a.published = ' . (int) $config['published']);
+					$query->where($db->quoteName('a.published') . ' = :published')
+						->bind(':published', $config['published'], ParameterType::INTEGER);
 				}
 				elseif ($config['published'] === '')
 				{
-					$query->where('a.published IN (0,1)');
+					$query->where($db->quoteName('a.published') . ' IN (0,1)');
```
Comment (Quy) [inline libraries/cms/html/menu.php:137]: Use `whereIn`

Comment (SharkyKZ) [reply libraries/cms/html/menu.php:137]: Do we really need this when using hardcoded values?

Comment (Quy) [reply libraries/cms/html/menu.php:137]: Maybe for consistency? Found 2 instances:

https://github.com/joomla/joomla-cms/blob/4.0-dev/administrator/components/com_csp/Model/ReportsModel.php#L119

https://github.com/joomla/joomla-cms/blob/4.0-dev/administrator/components/com_plugins/Model/PluginsModel.php#L250

Comment (HLeithner) [reply libraries/cms/html/menu.php:137]: We don't do it for hardcoded integer values in normal where statements so I think we can handle this the same. 

As long as all supported database engines can handle it.

File: libraries/cms/html/user.php [hunk 4]
```diff
@@ -13,6 +13,7 @@
 use Joomla\CMS\Factory;
 use Joomla\CMS\Helper\UserGroupsHelper;
 use Joomla\CMS\HTML\HTMLHelper;
+use Joomla\Database\ParameterType;
```
Comment (Quy) [inline libraries/cms/html/user.php]:
```suggestion
```

File: libraries/cms/html/access.php [hunk 5]
```diff
@@ -49,9 +49,9 @@ public static function level($name, $selected, $attribs = '', $params = true, $i
 	{
 		$db = Factory::getDbo();
 		$query = $db->getQuery(true)
-			->select($db->quoteName('a.id', 'value') . ', ' . $db->quoteName('a.title', 'text'))
+			->select($db->quoteName(['a.id', 'a.title'], ['value', 'text']))
```
Comment (Quy) [inline libraries/cms/html/access.php]: Will you rewrite in this format?
```
->select(
        [
                $db->quoteName('a.id', 'value'),
                $db->quoteName('a.title', 'text'),
        ]
)
```

Comment (SharkyKZ) [reply libraries/cms/html/access.php]: OK.

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 78a4901da2db10a5b59f02fc1adf48be80ba257c<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25948">issues.joomla.org/tracker/joomla-cms/25948</a>.</sub>

**Comment by anibalsanchez** [comment]
I have tested this item :white_check_mark: successfully on 78a4901da2db10a5b59f02fc1adf48be80ba257c<br /><br />From #JMAD19 PBF, Test OK<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25948">issues.joomla.org/tracker/joomla-cms/25948</a>.</sub>

**Comment by alikon** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25948">issues.joomla.org/tracker/joomla-cms/25948</a>.</sub>

**Comment by wilsonge** [comment]
Thanks!

**Pull request merged**